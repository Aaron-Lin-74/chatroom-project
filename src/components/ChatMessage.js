import { auth } from '../firebase'
import { useRef } from 'react'

function ChatMessage(props) {
  const { text, uid, photoURL, imageUrl, createdAt } = props.message
  const myModal = useRef(null)
  const imageModal = useRef(null)

  // set current user's message class to sent to distinguish from others
  const messageClass = uid === auth.currentUser.uid ? 'sent' : 'received'

  const displayTime = (date) => {
    // If in the same day, just show the hour:min, otherwise show the date and time
    const now = new Date()
    if (isSameDay(date, now)) {
      return formatTime(date.getHours(), date.getMinutes())
    }
    return `${date.toDateString()} ${formatTime(
      date.getHours(),
      date.getMinutes()
    )}`
  }

  const formatTime = (hour, minute) => {
    return `${hour < 10 ? '0' + hour : hour}:${
      minute < 10 ? '0' + minute : minute
    }`
  }
  // return true is two dates are of the same day
  const isSameDay = (date1, date2) => {
    if (
      date1.getFullYear() === date2.getFullYear() &&
      date1.getMonth() === date2.getMonth() &&
      date1.getDate() === date2.getDate()
    ) {
      return true
    }
    return false
  }

  const showModal = () => {
    // Get the image and insert it inside the modal
    myModal.current.style.display = 'block'
    imageModal.current.src = imageUrl
  }

  // When the user clicks on <span> (x), close the modal
  const closeModal = () => {
    // Get the <span> element that closes the modal
    myModal.current.style.display = 'none'
  }
  return (
    <>
      <div className='timeStamp'>
        <p className='messageTime'>
          {createdAt && displayTime(createdAt.toDate())}
        </p>
      </div>
      <div className={`message ${messageClass}`}>
        <img
          className='profilePhoto'
          src={photoURL || '/images/default_profile.jpg'}
          referrerPolicy='noreferrer'
          alt='profile'
        />
        {text ? (
          <p className='messageText'>{text}</p>
        ) : (
          <>
            <img
              className='messagePhoto'
              src={imageUrl}
              referrerPolicy='noreferrer'
              alt='messagePhoto'
              onClick={showModal}
            />
            <div className='modal' ref={myModal}>
              <span onClick={closeModal} className='closeModal'>
                &times;
              </span>
              <img
                className='modal-photo'
                referrerPolicy='noreferrer'
                alt='messagePhoto'
                ref={imageModal}
              />
            </div>
          </>
        )}
      </div>
    </>
  )
}
export default ChatMessage
