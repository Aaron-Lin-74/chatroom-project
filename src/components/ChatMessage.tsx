import { auth } from '../firebase';
import { useRef } from 'react';
import { MessageType } from '../types/message';

interface PropsType {
  message: MessageType;
  scrollDown: () => void;
}
function ChatMessage(props: PropsType) {
  const { text, uid, photoURL, imageUrl, createdAt } = props.message;
  const { scrollDown } = props;
  const myModal = useRef<HTMLDivElement>(null!);
  const imageModal = useRef<HTMLImageElement>(null!);
  const profileModal = useRef<HTMLDivElement>(null!);
  const profilePhoto = useRef<HTMLImageElement>(null!);

  // set current user's message class to sent to distinguish from others
  const messageClass = uid === auth.currentUser?.uid ? 'sent' : 'received';

  const showModal = () => {
    // Get the image and insert it inside the modal
    myModal.current.style.display = 'block';
    imageModal.current.setAttribute('src', imageUrl!);
  };

  // When the user clicks on <span> (x), close the modal
  const closeModal = () => {
    // Get the <span> element that closes the modal
    myModal.current.style.display = 'none';
    imageModal.current.setAttribute('src', '');
  };

  const showProfilePhoto = () => {
    profileModal.current.style.display = 'block';
    profilePhoto.current.src = photoURL || '/images/default_profile.jpg';
  };

  const closeProfileModal = () => {
    profileModal.current.style.display = 'none';
    profilePhoto.current.src = '';
  };

  return (
    <>
      <div className='timeStamp'>
        <p className='messageTime'>{createdAt}</p>
      </div>
      <div className={`message ${messageClass}`}>
        <img
          className='profilePhoto'
          src={photoURL || '/images/default_profile.jpg'}
          referrerPolicy='no-referrer'
          alt='profile'
          onClick={showProfilePhoto}
        />
        <div className='profile-modal' ref={profileModal}>
          <div className='profile-photo-container'>
            <span onClick={closeProfileModal} className='close-profile-modal'>
              &times;
            </span>
            <img
              className='big-profile-Photo'
              referrerPolicy='no-referrer'
              alt='profile'
              ref={profilePhoto}
            />
          </div>
        </div>
        {text ? (
          <p className='messageText'>{text}</p>
        ) : (
          <>
            <img
              className='messagePhoto'
              src={imageUrl}
              referrerPolicy='no-referrer'
              alt='messagePhoto'
              onClick={showModal}
              onLoad={scrollDown}
            />
            <div className='modal' ref={myModal}>
              <span onClick={closeModal} className='closeModal'>
                &times;
              </span>
              <img
                className='modal-photo'
                referrerPolicy='no-referrer'
                alt='messagePhoto'
                ref={imageModal}
              />
            </div>
          </>
        )}
      </div>
    </>
  );
}
export default ChatMessage;
