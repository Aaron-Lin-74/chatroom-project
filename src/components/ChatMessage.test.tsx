import { screen, render, waitFor, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import ChatMessage from './ChatMessage';

const fakeTextMessage = {
  id: '1234abcd',
  text: 'Hello World',
  uid: 'abc123',
  photoURL: 'http://www.fakePhotoURL',
  createdAt: '2020-10-10',
};

const fakePhotoMessage = {
  id: '1234abcd',
  uid: 'abc123',
  photoURL: 'http://www.fakePhotoURL',
  imageURL: 'http://www.fakeImageURL',
  createdAt: '2020-10-10',
};
const fakeScrollDown = jest.fn();

describe('ChatMessage component', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(
      <ChatMessage message={fakeTextMessage} scrollDown={fakeScrollDown} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render the component with text message', () => {
    render(
      <ChatMessage message={fakeTextMessage} scrollDown={fakeScrollDown} />
    );
    screen.getByText(fakeTextMessage.createdAt);
    screen.getByText(fakeTextMessage.text);
    screen.getByAltText('profile');
    screen.getByAltText('profile-modal');
  });

  test('should render the component with image message', () => {
    render(
      <ChatMessage message={fakePhotoMessage} scrollDown={fakeScrollDown} />
    );
    screen.getByText(fakePhotoMessage.createdAt);
    screen.getByAltText('profile');
    screen.getByAltText('profile-modal');
    screen.getByAltText('photoMessage');
    screen.getByAltText('photoMessage-modal');
  });

  test('should open image message modal when click the image message', () => {
    render(
      <ChatMessage message={fakePhotoMessage} scrollDown={fakeScrollDown} />
    );
    userEvent.click(screen.getByAltText('photoMessage'));
    expect(screen.getByTestId('photoMessage-modal')).toHaveStyle(
      'display: block'
    );
  });

  test('should close image message modal when click the × button', () => {
    render(
      <ChatMessage message={fakePhotoMessage} scrollDown={fakeScrollDown} />
    );
    userEvent.click(screen.getByAltText('photoMessage'));
    expect(screen.getByTestId('photoMessage-modal')).toHaveStyle(
      'display: block'
    );
    // The close button for photo message modal, the second x in this component
    userEvent.click(screen.getAllByText('×')[1]);
    expect(screen.getByTestId('photoMessage-modal')).toHaveStyle(
      'display: none'
    );
  });

  test('should open profile modal when click the profile image', () => {
    render(
      <ChatMessage message={fakePhotoMessage} scrollDown={fakeScrollDown} />
    );
    userEvent.click(screen.getByAltText('profile'));
    expect(screen.getByTestId('profile-modal')).toHaveStyle('display: block');
  });

  test('should close profile modal when click the × button', () => {
    render(
      <ChatMessage message={fakePhotoMessage} scrollDown={fakeScrollDown} />
    );
    userEvent.click(screen.getByAltText('profile'));
    expect(screen.getByTestId('profile-modal')).toHaveStyle('display: block');
    // The close button for photo message modal, the first x in this component
    userEvent.click(screen.getAllByText('×')[0]);
    expect(screen.getByTestId('profile-modal')).toHaveStyle('display: none');
  });

  test('should call the scrollDown function when the photo message loaded', () => {
    render(
      <ChatMessage message={fakePhotoMessage} scrollDown={fakeScrollDown} />
    );
    fireEvent.load(screen.getByAltText('photoMessage'));
    expect(fakeScrollDown).toBeCalledTimes(1);
  });
});
