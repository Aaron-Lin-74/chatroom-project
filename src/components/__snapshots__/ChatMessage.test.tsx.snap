// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessage component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="timeStamp"
  >
    <p
      class="messageTime"
    >
      2020-10-10
    </p>
  </div>
  <div
    class="message received"
  >
    <img
      alt="profile"
      class="profilePhoto"
      referrerpolicy="no-referrer"
      src="http://www.fakePhotoURL"
    />
    <div
      class="profile-modal"
      data-testid="profile-modal"
    >
      <div
        class="profile-photo-container"
      >
        <span
          class="close-profile-modal"
        >
          Ã—
        </span>
        <img
          alt="profile-modal"
          class="big-profile-Photo"
          referrerpolicy="no-referrer"
        />
      </div>
    </div>
    <p
      class="messageText"
    >
      Hello World
    </p>
  </div>
</DocumentFragment>
`;
